src/components/popup/popup.jsx

```diff
@@ -0,0 +1,266 @@
+'use strict';
+
+// TODO
+// - support block props
+//   - zIndexGroupLevel
+// - nested popups
+// - hide popup if target isn't visible
+// - handle window resize - do we really need it?
+// - tests
+
+const React = require('react');
+const ReactDOM = require('react-dom');
+
+const BEM = require('../../core/bem/bem');
+const provide = require('../../provider/provider');
+
+const ZINDEX_FACTOR = 1000;
+const visiblePopupsZIndexes = {};
+
+const RenderInBody = require('../render-in-body/render-in-body.jsx');
+const calcBestDrawingParams = require('./calc-drawing-params').calcBestDrawingParams;
+const isEventOutsideClientBounds = require('../../lib/window').isEventOutsideClientBounds;
+
+module.exports = class Popup extends BEM {
+
+    constructor(props) {
+        super(props);
+
+        // TODO
+        // this._parentPopup = undefined;
+        // this._zIndex = null;
+        // this._zIndexGroupLevel = null;
+
+        this.domElementPopup = null;
+        this.anchor = null;
+        this.position = null;
+        this.clickEventBindTimeout = null;
+        this.isWindowClickBinded = false;
+
+        this.defaultProps = {
+            visible: false,
+            autoclosable: false,
+            target: 'anchor',
+            size: 's'
+        };
+
+        this.state = {
+            direction: null,
+            hovered: false,
+            styles: {
+                top: 0,
+                left: 0
+            }
+        };
+
+        this.handleMouseEnter = this.handleMouseEnter.bind(this);
+        this.handleMouseLeave = this.handleMouseLeave.bind(this);
+        this.handleWindowClick = this.handleWindowClick.bind(this);
+
+        this.redraw = this.redraw.bind(this);
+    }
+
+    componentWillReceiveProps(nextProps) {
+        if (nextProps.visible) {
+            this.redraw();
+        }
+    }
+
+    componentDidMount() {
+        this.props.autoclosable && this.ensureClickEvent();
+    }
+
+    componentWillUnmount() {
+        this.props.autoclosable && this.ensureClickEvent(true);
+    }
+
+    componentDidUpdate(prevProps, prevState) {
+        if (this.props.autoclosable) {
+            if (prevProps.onClickOutside !== this.props.onClickOutside) {
+                this.ensureClickEvent();
+            } else {
+                if (prevProps.visible !== this.props.visible) {
+                    this.ensureClickEvent(!this.props.visible);
+                }
+            }
+        }
+    }
+
+    render() {
+        return (
+            <RenderInBody>
```

Comment:
We should rewrite it. We don't need the new component. Its functions should be in the popup.